server/src/main/java/org/elasticsearch/action/admin/cluster/node/reload/NodesReloadSecureSettingsRequest.java

```diff
@@ -62,24 +68,60 @@ public ActionRequestValidationException validate() {
         return validationException;
     }
 
-    public String secureSettingsPassword() {
+    public SecureString secureSettingsPassword() {
         return secureSettingsPassword;
     }
 
-    public NodesReloadSecureSettingsRequest secureStorePassword(String secureStorePassword) {
+    public NodesReloadSecureSettingsRequest secureStorePassword(SecureString secureStorePassword) {
         this.secureSettingsPassword = secureStorePassword;
         return this;
     }
 
     @Override
     public void readFrom(StreamInput in) throws IOException {
         super.readFrom(in);
-        secureSettingsPassword = in.readString();
+        final byte[] passwordBytes = in.readByteArray();
+        try {
+            this.secureSettingsPassword = new SecureString(utf8BytesToChars(passwordBytes));
+        } finally {
+            Arrays.fill(passwordBytes, (byte) 0);
+        }
     }
 
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         super.writeTo(out);
-        out.writeString(secureSettingsPassword);
+        final byte[] passwordBytes = toUtf8Bytes(this.secureSettingsPassword.getChars());
+        try {
+            out.writeByteArray(passwordBytes);
+        } finally {
+            Arrays.fill(passwordBytes, (byte) 0);
+        }
+    }
+
+    /**
+     * Encodes the provided char[] to a UTF-8 byte[]. The provided char[] is not
+     * modified by this method, so the caller needs to take care of clearing the
+     * value if it is sensitive.
+     */
+    private static byte[] toUtf8Bytes(char[] chars) {
+        final CharBuffer charBuffer = CharBuffer.wrap(chars);
+        final ByteBuffer byteBuffer = StandardCharsets.UTF_8.encode(charBuffer);
+        final byte[] bytes = Arrays.copyOfRange(byteBuffer.array(), byteBuffer.position(), byteBuffer.limit());
```

Comment:
This code is dangerous so I think we should correct it. The char buffer is guaranteed to be a HeapCharBuffer by the javadocs; however there is no such guarantee about the ByteBuffer returned by `StandardCharsets.UTF_8.encode(charBuffer)`. We need to account for whether the bytebuffer has an array that can be accessed or not.

```java
final byte[] bytes;
if (byteBuffer.hasArray()) {
    // there is no guarantee that the byte buffers backing array is the right size so we need to make a copy
    bytes = Arrays.copyOfRange(byteBuffer.array(), byteBuffer.position(), byteBuffer.limit());
    Arrays.fill(byteBuffer.array(), (byte) 0); // clear sensitive data
} else {
    final int length = byteBuffer.limit() - byteBuffer.position();
    bytes = new byte[length];
    byteBuffer.get(bytes);
    // if the buffer is not read only we can reset and fill with 0's
    if (byteBuffer.isReadOnly() == false) {
         byteBuffer.clear(); // reset
         for (int i = 0; i < byteBuffer.limit(); i++) {
             byteBuffer.put((byte) 0);
         }
    }
}
```

Reply:
Thanks for the detailed snippet.